docs/microbenchmark-design-guidelines.md

```diff
@@ -0,0 +1,557 @@
+
+### IterationSetup
+
+If your benchmark requires a clean state for every invocation, you need to use the `[IterationSetup]` attribute. Unfortunately, just using the `[IterationSetup]` attribute is not enough to get stable results. You also need to make sure that the benchmark itself performs enough of computations for a single invocation to run longer than 100ms. The closer to 250ms the more stable results you are going to get.
+
+```cs
+[Params(1000 * 1000 * 200)] // allows for stable iteraiton around 200ms
+public int NumberOfBytes { get; set; }
+
+private byte[] _source, _destination;
+
+[GlobalSetup]
+public void Setup()
+{
+    _source = ValuesGenerator.Array<byte>(NumberOfBytes);
+    _destination = new byte[Base64.GetMaxEncodedToUtf8Length(NumberOfBytes)];
+}
+
+[IterationSetup(Target = nameof(Base64EncodeInPlace))]
+public void SetupBase64EncodeInPlace() => Array.Copy(_source, _destination, _source.Length);
+
+[Benchmark]
+public OperationStatus Base64EncodeInPlace() => Base64.EncodeToUtf8InPlace(_destination, _source.Length, out _);
+```
+
+**Note:** If you need to clean up resources after every benchmark run, you should use the corresponding `[IterationCleanup]` attribute. It's going to be executed once after every iteration.
+
+If you want to get a better understanding of it, you should read [this blog post](https://aakinshin.net/posts/stopwatch/#pitfalls) about stopwatch and follow the GitHub discussions in this [issue](https://github.com/dotnet/BenchmarkDotNet/issues/730) and [PR](https://github.com/dotnet/BenchmarkDotNet/pull/760).
+
+**Note:** if using `[GlobalSetup]` is enough, you should not be using `[IterationSetup]`.
+
+### OperationsPerInvoke
+
+`[GlobalSetup]` and `[IterationSetup]` might not be enough if you want to setup some nano-benchmarks.
+
+A good example is `Slicing` a `Span`. `Span` is a stack-only type, so we can not have a `[GlobalSetup]` method which writes it to a field of a `class`.
+
+```cs
+class WontCompile
+{
+    Span<byte> _span; // compiler error
+
+    [GlobalSetup]
+    public void Setup() => _span = new Span<byte>(new byte[10]);
+}
+```
+
+The current implementation of BenchmarkDotNet does not allow to run benchmarks defined in structs (including `ref struct`).
+
+If we write a following benchmark:
+
+```cs
+private byte[] _nonEmptyArray = new byte[10];
+
+[Benchmark]
+public Span<byte> Slice()
+{
+    Span<byte> span = new Span<byte>(_nonEmptyArray);
+
+    return span.Slice(span.Length / 2);
+}
+```
+
+The benchmark is not going to measure the performance of `Slice` operation, but similar to previous example it's going to measure the performance of creating a `Span` from `Array` and the `Slice` operation.
+
+To solve this problem we can use `OperationsPerInvoke` property of the `[Benchmark]` attribute.
+
+```cs
+private byte[] _nonEmptyArray = new byte[18];
+
+[Benchmark(OperationsPerInvoke = 16)]
```

Comment:
Is there any guidance or principle around how I should choose this number? Will BDN scale the number of invocations with this, since we'll hit the 250ms ceiling still?

Reply:
Yes, it's going to scale the number of invocations.

I added following sentence to the description:

**Note:** `OperationsPerInvoke` should be big enough to amortize the "setup" cost.